Northstar.Custom/mod/scripts/vscripts/_testing.nut

```diff
@@ -0,0 +1,302 @@
+	}
+	catch ( exception )
+	{
+		test.completed = false
+		test.error = exception
+	}
+}
+
+void function PrintAllTestResults()
+{
+	int totalSucceeded = 0
+	int totalFailed = 0
+	int totalErrored = 0
+
+	foreach ( category, categoryTests in file.tests )
+	{
+		int categorySucceeded = 0
+		int categoryFailed = 0
+		int categoryErrored = 0
+
+		printt( format( "Results for category: '%s'", category ) )
+		foreach ( test in categoryTests )
+		{
+			if ( test.completed )
+			{
+				if ( test.passed )
+				{
+					printt( "\t", test.testName, "- Passed!" )
+					categorySucceeded++
+				}
+				else
+				{
+					printt( "\t", test.testName, "- Failed!" )
+					printt( "\t\tExpected:", test.expectedResult )
+					printt( "\t\tActual:  ", test.actualResult )
+					categoryFailed++
+				}
+			}
+			else
+			{
+				printt( "\t", test.testName, "- Errored!" )
+				printt( "\t\tError:", test.error )
+				categoryErrored++
+			}
+		}
+
+		printt( "Succeeded:", categorySucceeded, "Failed:", categoryFailed, "Errored:", categoryErrored )
+
+		totalSucceeded += categorySucceeded
+		totalFailed += categoryFailed
+		totalErrored += categoryErrored
+	}
+
+	printt( "TOTAL SUCCEEDED:", totalSucceeded, "TOTAL FAILED:", totalFailed, "TOTAL ERRORED:", totalErrored )
+}
+
+void function PrintCategoryResults( string category )
+{
+	int categorySucceeded = 0
+	int categoryFailed = 0
+	int categoryErrored = 0
+
+	printt( format( "Results for category: '%s'", category ) )
+	foreach ( test in file.tests[ category ] )
+	{
+		if ( test.completed )
+		{
+			if ( test.passed )
+			{
+				printt( "\t", test.testName, "- Passed!" )
+				categorySucceeded++
+			}
+			else
+			{
+				printt( "\t", test.testName, "- Failed!" )
+				printt( "\t\tExpected:", test.expectedResult )
+				printt( "\t\tActual:  ", test.actualResult )
+				categoryFailed++
+			}
+		}
+		else
+		{
+			printt( "\t", test.testName, "- Errored!" )
+			printt( "\t\tError:", test.error )
+			categoryErrored++
+		}
+	}
+
+	printt( "Succeeded:", categorySucceeded, "Failed:", categoryFailed, "Errored:", categoryErrored )
+}
+
+void function PrintTestResult( TestInfo test )
+{
+	string resultString = test.testName
+
+	if ( test.completed )
+	{
+		if ( test.passed )
+			resultString += " - Passed!"
+		else
+		{
+			resultString += " - Failed!"
+			resultString += "\n\tExpected: " + test.expectedResult
+			resultString += "\n\tActual:   " + test.actualResult
+		}
+	}
+	else
+	{
+		resultString += " - Not completed!"
+		resultString += "\n\tError: " + test.error
+	}
+
+	printt( resultString )
+}
+
+void function AddTest( string testCategory, string testName, var functionref() testFunc, var expectedResult )
+{
+	TestInfo newTest
+	newTest.testName = testName
+	newTest.callback = testFunc
+	newTest.expectedResult = expectedResult
+
+	// create the test category if it doesn't exist
+	if ( !( testCategory in file.tests ) )
+		file.tests[ testCategory ] <- [ newTest ]
+	else
+		file.tests[ testCategory ].append( newTest )
+}
```

Comment:
Missing newline ^^"

(I can't add that via GitHub suggestion cause for some stupid reason that always adds two new lines on last line x_X)

Reply:
Fixed in 5414c3fdad5ed86e82d0d925a064a7d3263c7b53